### PR TITLE
feat(pruning): prune ChangeSets & History during pipeline

### LINKED ISSUE #3433: Respect `Account History` prune part in the pipeline
### Describe the feature

We need to start writing data related to the `Account History` prune part during the initial sync only when the pruning target is reached:
- `ExecutionStage` writes to the `AccountChangeSet` table
- `IndexAccountHistoryStage` writes to the `AccountHistory` table

### Additional context

_No response_

### LINKED ISSUE #3434: Respect `Storage History` prune part in the pipeline
### Describe the feature

We need to start writing tables related to the `Storage History` prune part during the initial sync only when the pruning target is reached:
- `ExecutionStage` writes to the `StorageChangeSet` table
- `IndexStorageHistoryStage` writes to the `StorageHistory` table

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by joshieDo**
Resolves https://github.com/paradigmxyz/reth/issues/3433
Resolves https://github.com/paradigmxyz/reth/issues/3434

* Adds account and storage changeset pruning during execution stage (if set). If so, it will also prune during the `AccountHistory` and `StorageHistory` stages. 
* If pruning is enabled, hashing and merkle stages might have to be run from scratch on subsequent pipeline runs if there are not enough changesets. ([more context](https://github.com/paradigmxyz/reth/pull/3728#issuecomment-1655470806))

Full pruning *is not* supported, and will throw error at reading the configuration. Minimum 64 block distance.

### CODE REVIEW

**Comment by joshieDo** [comment]
since there was some overlap, should https://github.com/paradigmxyz/reth/pull/3733 be pushed first

File: crates/stages/src/stages/index_account_history.rs [hunk 1]
```diff
@@ -12,18 +15,20 @@ pub struct IndexAccountHistoryStage {
     /// Number of blocks after which the control
     /// flow will be returned to the pipeline for commit.
     pub commit_threshold: u64,
+    /// Pruning configuration.
+    pub prune_targets: PruneModes,
```
Comment (shekhirin) [inline crates/stages/src/stages/index_account_history.rs:19]:
```suggestion
    pub prune_modes: PruneModes,
```

File: crates/stages/src/stages/index_account_history.rs [hunk 2]
```diff
@@ -38,12 +43,20 @@ impl<DB: Database> Stage<DB> for IndexAccountHistoryStage {
     async fn execute(
         &mut self,
         provider: &DatabaseProviderRW<'_, &DB>,
-        input: ExecInput,
+        mut input: ExecInput,
```
Comment (shekhirin) [inline crates/stages/src/stages/index_account_history.rs:46]: we're modifying the copy of the `Input` and not the mutable reference, so it should be fine

File: crates/stages/src/stages/mod.rs [hunk 3]
```diff
@@ -35,3 +35,186 @@ pub use merkle::*;
...
+        provider.commit().unwrap();
+
+        let check_pruning = |factory: Arc<ProviderFactory<_>>,
+                             prune_targets: PruneModes,
+                             expect_num_receipts: usize,
+                             expect_num_acc_changesets: usize,
+                             expect_num_storage_changesets: usize| async move {
+            let provider: DatabaseProviderRW<'_, &DatabaseEnv> = factory.provider_rw().unwrap();
+
+            let mut execution_stage = ExecutionStage::new(
```
Comment (shekhirin) [inline crates/stages/src/stages/mod.rs:132]:
```suggestion
            // Check Execution
            let mut execution_stage = ExecutionStage::new(
```

File: crates/stages/src/stages/index_account_history.rs [hunk 4]
```diff
@@ -38,12 +43,20 @@ impl<DB: Database> Stage<DB> for IndexAccountHistoryStage {
...
     ) -> Result<ExecOutput, StageError> {
         if input.target_reached() {
             return Ok(ExecOutput::done(input.checkpoint()))
         }
```
Comment (shekhirin) [inline crates/stages/src/stages/index_account_history.rs:50]: should we move this check after the `input.checkpoint` adjustment?

**Review by shekhirin — APPROVED** [review]
LGTM overall, nits. Appreciate the tests!

**Comment by joshieDo** [comment]
With this PR, lower end machines might not be able to catch up if their pruning is very aggressive. This happens because the pipeline re-run that happens after the first sync will trigger the hashing and merkle stages to be run from scratch since there are not enough changesets.

Follow-up PR will handle that, since it requires a few too many changes for this PR

**Review by mattsse — APPROVED** [review]
nit re duplicate condition,

otherwise lgtm

File: crates/primitives/src/prune/part.rs [hunk 5]
```diff
@@ -1,8 +1,10 @@
+use derive_more::Display;
 use reth_codecs::{main_codec, Compact};
+use thiserror::Error;
 
 /// Part of the data that can be pruned.
 #[main_codec]
-#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Debug, Display, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
 pub enum PrunePart {
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply crates/primitives/src/prune/part.rs:8]: Personally, "Step" sounds like a consecutive action like "Stage" in the context of pipeline, while in pruning you can enable different parts and it doesn't really matter in which order and composition they are executed. But I do agree that "Part" is weird too.

File: crates/primitives/src/prune/part.rs [hunk 6]
```diff
@@ -1,8 +1,10 @@
+use derive_more::Display;
 use reth_codecs::{main_codec, Compact};
+use thiserror::Error;
 
 /// Part of the data that can be pruned.
 #[main_codec]
-#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Debug, Display, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
```
[the remark this answers is not in the document]
Comment (joshieDo) [reply crates/primitives/src/prune/part.rs:7]: this is coming from `derive_more`, which I noticed being used somewhere else in the crate

**Pull request merged**